lib/shelley/src/Cardano/Wallet/Shelley/Network/Blockfrost.hs

```diff
@@ -287,24 +400,382 @@ withNetworkLayer tr network np project k =
+  utxos <- BF.getTxUtxos hash
+  withdrawals <- BF.getTxWithdrawals hash
+  metadata <- BF.getTxMetadataJSON hash
+  (transaction,) <$>
+    assembleTransaction nd transaction utxos withdrawals metadata
+
+assembleTransaction
+    :: forall m
+     . MonadError BlockfrostError m
+    => SomeNetworkDiscriminant
+    -> BF.Transaction
+    -> BF.TransactionUtxos
+    -> [BF.TransactionWithdrawal]
+    -> [BF.TransactionMetaJSON]
+    -> m Tx
+assembleTransaction
+    (SomeNetworkDiscriminant (Proxy :: Proxy nd))
+    BF.Transaction{..}
+    BF.TransactionUtxos{..}
+    txWithdrawals
+    metadataJSON = liftEither do
+        txId <- parseTxHash _transactionHash
+        let fee = Just $ Coin $ fromIntegral _transactionFees
+        (resolvedInputs, resolvedCollateralInputs) <-
+            fromInputs _transactionUtxosInputs
+        outputs <- for _transactionUtxosOutputs \out@BF.UtxoOutput{..} -> do
+            let outAddr = BF.unAddress _utxoOutputAddress
+            address <- either (throwError . InvalidAddress outAddr) pure $
+                decodeAddress @nd outAddr
+            tokens <- do
+                coin <- case [ lovelaces
+                             | BF.AdaAmount lovelaces <- _utxoOutputAmount ] of
+                    [l] -> fromBlockfrost l
+                    _ -> throwError $ InvalidUtxoOutputAmount out
+                pure $ TokenBundle coin mempty -- TODO: Handle native assets
+            pure TxOut{..}
+        withdrawals <- Map.fromList <$>
+            for txWithdrawals ( \BF.TransactionWithdrawal{..} -> do
+                let addr = BF.unAddress _transactionWithdrawalAddress
+                rewardAccount <-
+                    first (InvalidAddress addr) $ decodeStakeAddress @nd addr
+                coin <- fromBlockfrost _transactionWithdrawalAmount
+                pure (rewardAccount, coin)
+            )
+        metadata <-
+            if null metadataJSON
+            then pure Nothing
+            else Just . TxMetadata . Map.fromList . catMaybes <$>
+                for metadataJSON ( \BF.TransactionMetaJSON{..} -> do
+                    label <- either (throwError . InvalidTxMetadataLabel) pure $
+                        readEither (T.unpack _transactionMetaJSONLabel)
+                    fmap (label,) <$> for _transactionMetaJSONJSONMetadata
+                        (first InvalidTxMetadataValue . unmarshalMetadataValue)
+                    )
+        let scriptValidity = Just $
+                if _transactionValidContract
+                    then TxScriptValid
+                    else TxScriptInvalid
+        pure Tx
+            { txId
+            , fee
+            , resolvedInputs
+            , resolvedCollateralInputs
+            , outputs
+            , collateralOutput = Nothing
+            , withdrawals
+            , metadata
+            , scriptValidity
+            }
+
+  where
+    unmarshalMetadataValue :: Aeson.Value -> Either String TxMetadataValue
```

Comment:
👍 Solid reverse engineering! Perhaps make it top-level so that it's easier to test manually?

Reply:
:ok: